tidy3d/components/monitor.py

```diff
@@ -667,6 +667,69 @@ def storage_size(self, num_cells: int, tmesh: ArrayLike[float, 1]) -> int:
         return BYTES_COMPLEX * len(self.ux) * len(self.uy) * len(self.freqs) * 4
 
 
+class DiffractionEfficiencyMonitor(FieldMonitor, PlanarMonitor):
+    """:class:`Monitor` that uses a 2D Fourier transform to compute the
+    diffraction efficiency for given (or all) diffraction orders.
+
+    Example
+    -------
+    >>> monitor = DiffractionEfficiencyMonitor(
+    ...     center=(1,2,3),
+    ...     size=(5,5,0),
+    ...     freqs=[250e12, 300e12],
+    ...     name='diffraction_efficiency_monitor',
+    ...     normal_dir='+',
+    ...     orders_x=[0, 1, 2],
+    ...     orders_y=[0],
+    ...     )
+    """
+
+    normal_dir: Direction = pydantic.Field(
+        "+",
+        title="Normal vector orientation",
+        description="Direction of the surface monitor's normal vector w.r.t. "
+        "the positive x, y or z unit vectors. Must be one of ``'+'`` or ``'-'``. "
+        "Defaults to ``'+'`` if not provided.",
+    )
+
+    orders_x: Tuple[int, ...] = pydantic.Field(
+        None,
+        title="Diffraction orders along x",
+        description="Diffraction orders along x for which efficiency should be computed. "
+        "If ``None``, all orders between -1000 and 1000 will be considered.",
+    )
+
+    orders_y: Tuple[int, ...] = pydantic.Field(
+        None,
+        title="Diffraction orders along y",
+        description="Diffraction orders along y for which efficiency should be computed. "
+        "If ``None``, all orders between -1000 and 1000 will be considered.",
+    )
+
+    @pydantic.root_validator()
```

Comment:
let's just do this as a regular validator. One way would be to define in `validators.py`

```py
def validate_orders(x_or_y: str, max_orders:int=100):
    field_name = f'orders_{x_or_y}'
    @pydantic.validator(field_name, always=True)
    def _set_default_diffraction_orders(cls, val):
        if val is None:
            return np.arange(-100, 101).tolist()
        return val
    return _set_default_diffraction_orders
```

and then import this function in this module and insert in the main body (same location as this validator) as
```py
    _validate_orders_x = validate_orders('x', max_orders=MAX_ORDERS)
    _validate_orders_y = validate_orders('y', max_orders=MAX_ORDERS)
```